customize: Add documentation for configuration variables (#21)
inveniosoftware/docs-invenio-ils: docs/customize/configure.md

# Configuration Variables

> This document provides a comprehensive list and explanation of the configuration variables available in the `invenio-app-ils` backend. These variables can be set in the configuration file(s) to customize the behavior of your Invenio ILS application.

Reviewer: Same remark/comment as in the other PR. I would stick to `InvenioILS` (one word) and avoid modules name.
Or, if you want to explicit mention `invenio-app-ils`, then I would explain that is the module (and not really the backend in the true sense of the term) with a link:
```suggestion
This document provides a comprehensive list and explanation of the configuration variables available in the [invenio-app-ils](https://github.com/inveniosoftware/invenio-app-ils) module. These variables can be set in the configuration file(s) to customize the behavior of your InvenioILS application.
```

## REST Configurations

> ### `ILS_RECORDS_REST_ENDPOINTS`

Reviewer 2: I wouldn't add this one, I think it is too advanced, maybe we need some "advanced"  config section or remove completely

REST endpoints for various record types.

```python
ILS_RECORDS_REST_ENDPOINTS = {
    'books': {
        'pid_type': 'bookid',
        'path': '/books',
        'default_media_type': 'application/json',
    },
    'members': {
        'pid_type': 'memberid',
        'path': '/members',
        'default_media_type': 'application/json',
    },
}
```

### `ILS_MAILS_TEMPLATES`

Paths to email templates used for various notifications.

```python
ILS_MAILS_TEMPLATES = {
    'loan_due': 'invenio_app_ils/loans/email/loan_due.html',
    'loan_overdue': 'invenio_app_ils/loans/email/loan_overdue.html',
}
```

## Circulation Configuration

### `CIRCULATION_LOAN_TRANSITIONS`

A dictionary defining the possible transitions for loans.

```python

> CIRCULATION_LOAN_TRANSITIONS = {

Reviewer 2: this I would describe a bit more as a use case: for example if you need to add a loan in transit state or skip a step in the loan circulation, then you need to modify this variable. Also a bit advanced config.

    'CREATED': ['PENDING'],
    'PENDING': ['ITEM_ON_LOAN', 'CANCELLED'],
    'ITEM_ON_LOAN': ['RETURNED'],
    'RETURNED': [],
    'CANCELLED': [],
}
```

For detailed documentation for circulation sub-module, refer to [invenio-circulation](https://invenio-circulation.readthedocs.io/en/latest/).

## Debug Configuration

### DEBUG (True/False)

Enable or disable debugging mode.

> ### DEBUG_TB_ENABLED (True/False)

Reviewer 2: this one does not work, to remove

Enable or disable the debug toolbar.

> ### DEBUG_TB_INTERCEPT_REDIRECTS (True/False)

Reviewer 2: same as above

Enable or disable intercepting redirects in the debug toolbar.

## Rate Limiting Configuration

### RATELIMIT_STORAGE_URL

Storage URL for the rate limiter.

```python

> RATELIMIT_STORAGE_URL = "redis://localhost:6379/3"

Reviewer 2: this is more "for website administrators" section of configuration, not features related one

```

### RATELIMIT_AUTHENTICATED_USER

Rate limit for authenticated users.

```python
RATELIMIT_AUTHENTICATED_USER = "5000 per hour;150 per minute"
```

### RATELIMIT_GUEST_USER

Rate limit for non-authenticated users.

```python
RATELIMIT_GUEST_USER = "1000 per hour;100 per minute"
```

## Internationalization (I18N) Configuration

### BABEL_DEFAULT_LANGUAGE

Default language setting.

```python

> BABEL_DEFAULT_LANGUAGE = "en"

Reviewer 2: I would add a disclaimer that for now the software is available only in english

```

### BABEL_DEFAULT_TIMEZONE

Default timezone setting.

```python
BABEL_DEFAULT_TIMEZONE = "Europe/Zurich"
```

## Notifications Configuration

> ### ILS_NOTIFICATIONS_BACKENDS_BUILDER

Reviewer 2: this variable and the next are advanced config also

Factory function to use when sending notifications.

```python
ILS_NOTIFICATIONS_BACKENDS_BUILDER = (
    "invenio_app_ils.notifications.backends:notifications_backend_builder"
)
```

### ILS_NOTIFICATIONS_MSG_BUILDER

Notification message creator.

```python
ILS_NOTIFICATIONS_MSG_BUILDER = (
    "invenio_app_ils.notifications.messages:notification_msg_builder"
)
```

### ILS_NOTIFICATIONS_TEMPLATES

Override default global common templates.

```python

> ILS_NOTIFICATIONS_TEMPLATES = {"footer": "footer.html"}

Reviewer 2: I would give more possible key value pairs here to show what is possible

```

### ILS_NOTIFICATIONS_MSG_BUILDER_DOCUMENT_REQUEST

Notification message creator for document requests notifications.

```python

> ILS_NOTIFICATIONS_MSG_BUILDER_DOCUMENT_REQUEST = "invenio_app_ils.document_requests.notifications.messages:notification_document_request_msg_builder"

Reviewer 2: this one is also advanced config

```

### ILS_NOTIFICATIONS_TEMPLATES_DOCUMENT_REQUEST

Override default document requests templates.

```python
ILS_NOTIFICATIONS_TEMPLATES_DOCUMENT_REQUEST = {}
```

### ILS_NOTIFICATIONS_FILTER_DOCUMENT_REQUEST

> Function to select and filter which notifications should be sent.

Reviewer 2: ```suggestion
Function to select and filter which notifications should be sent. Possibility to filter out some emails, based on your needs.
```

```python
ILS_NOTIFICATIONS_FILTER_DOCUMENT_REQUEST = document_request_notification_filter
```

## Email Configuration

### SUPPORT_EMAIL

Email address for support.

```python
SUPPORT_EMAIL = "info@example.com"
```

### MAIL_SUPPRESS_SEND (True/False)

Disable email sending by default.

### MAIL_NOTIFY_SENDER

Email address for email notification sender.

```python
MAIL_NOTIFY_SENDER = "noreply@example.com"
```

### MAIL_NOTIFY_CC (list)

Email CC address(es) for email notifications.

### MAIL_NOTIFY_BCC (list)

Email BCC address(es) for email notifications.

### ILS_MAIL_ENABLE_TEST_RECIPIENTS (True/False)

Enable or disable sending mail to test recipients.

### ILS_MAIL_NOTIFY_TEST_RECIPIENTS (list)

When `ILS_MAIL_ENABLE_TEST_RECIPIENTS` is True, all emails are sent here.

```python
ILS_MAIL_NOTIFY_TEST_RECIPIENTS = ["john.doe@example.com"]
```

## Assets Configuration

> ### COLLECT_STORAGE

Reviewer 2: advanced configuration

Here's the continuation of the comprehensive documentation for the provided Python config file in Markdown format:

## Celery Configuration

### CELERY_BROKER_URL

URL for the Celery message broker (default: RabbitMQ).

### CELERY_RESULT_BACKEND

URL for the Celery result backend (default: Redis).

### CELERY_BEAT_SCHEDULE

Scheduled tasks configuration for Celery (cronjobs).

Example:

```python
CELERY_BEAT_SCHEDULE = {
    "indexer": {
        "task": "invenio_indexer.tasks.process_bulk_queue",
        "schedule": timedelta(minutes=5),
    },
    "send_overdue_loan_reminders": {
        "task": "invenio_app_ils.circulation.notifications.tasks.send_overdue_loans_notification_reminder",
        "schedule": crontab(minute=0, hour=2),
    },
    "stats-aggregate-events": {
        "task": "invenio_stats.tasks.aggregate_events",
        "schedule": timedelta(hours=3),
        "args": [("record-view-agg", "file-download-agg")],
    },
}
```

## Feature Configurations

### EXTEND_LOANS_LOCATION_UPDATED (True/False)

> Flag to indicate whether to start a task to update active loans when closures of a location are updated.

Reviewer 2: I think this needs a bit more context to explain what is happening

### EXTEND_LOANS_SCHEDULE_TIME

Time at which the task to extend loans should be run.

```python
EXTEND_LOANS_SCHEDULE_TIME = datetime.time(2)
```

### ILS_SELF_CHECKOUT_ENABLED (True/False)

> Enable or disable Self-Checkout feature endpoint.

Reviewer 2: ```suggestion
Enable or disable Self-Checkout feature.
```

## Database Configuration

### SQLALCHEMY_DATABASE_URI

Database URI including user and password.

> ```python

Reviewer 2: I think it should be clustered with other website admin/developer variables

SQLALCHEMY_DATABASE_URI = "postgresql+psycopg2://test:psw@localhost/ils"
```

## JSONSchemas Configuration

### JSONSCHEMAS_HOST

Hostname used in URLs for local JSONSchemas.

```python
JSONSCHEMAS_HOST = "127.0.0.1:5000"
```

## CSRF Configuration

### REST_CSRF_ENABLED (True/False)

Enable or disable CSRF protection for REST APIs.

## CORS Configuration

### REST_ENABLE_CORS (True/False)

Enable or disable CORS support for REST APIs.

### CORS_SEND_WILDCARD (True/False)

Allow or disallow credentials to be sent with CORS requests.

### CORS_SUPPORTS_CREDENTIALS (True/False)

> Exable or disable supporting credentials for CORS requests.

Reviewer 2: these are "security variables", we should add a warning that it is not advised to modify them

## Flask Configuration

> ### SECRET_KEY

Reviewer 2: this is also security variable, we need a note indicating that this needs to be changed to a secure key when website goes to production

Secret key used for session management and CSRF protection.

### MAX_CONTENT_LENGTH

Maximum allowed size for uploaded files via form data.

```python
MAX_CONTENT_LENGTH = 100 * 1024 * 1024  # 100 MiB
```

### SESSION_COOKIE_SECURE (True/False)

Set cookie with the secure flag by default.

### SESSION_COOKIE_SAMESITE

Set SameSite attribute for session cookies.

```python
SESSION_COOKIE_SAMESITE = "Lax"

> ```

Reviewer 2: also "security related" variables

### APP_ALLOWED_HOSTS

List of allowed hosts for the application.

```python
APP_ALLOWED_HOSTS = ["localhost", "127.0.0.1"]
```

### SPA_HOST

Host URL for Single Page Application.

```python
SPA_HOST = "https://127.0.0.1:3000"
```

> ### SPA_PATHS

Reviewer 2: candidates for site admin/devops variables section

Paths and placeholders for Single Page Application routes.

```python

> SPA_PATHS = dict(

Reviewer 2: I would drop this one from docs, I am not sure what will happen if it is modified

    literature="/literature/%(pid)s",
    loan="/backoffice/loans",
    patron="/backoffice/patrons",
    profile="/profile",
    search="/search?q=%(querystring)s",
)
```

Here's the continuation and completion of the configuration file documentation in Markdown format:

## Records REST Configuration

### RECORDS_REST_MAX_RESULT_WINDOW

Maximum result window for Records REST endpoints.

```python
RECORDS_REST_MAX_RESULT_WINDOW = 10000
```

### RECORDS_REST_VOCAB_MAX_RESULT_WINDOW

Maximum result window for vocabulary Records REST endpoints.

```python
RECORDS_REST_VOCAB_MAX_RESULT_WINDOW = 500
```

### PIDSTORE_RECID_FIELD

Field name in the PID store for the record identifier.

```python

> PIDSTORE_RECID_FIELD = "pid"

Reviewer 2: I would drop this one

```

### REST_MIMETYPE_QUERY_ARG_NAME

Name of the URL argument to choose response serializer.

```python

> REST_MIMETYPE_QUERY_ARG_NAME = "format"

Reviewer 2: I would drop this one as well

```

### RECORDS_REST_ENDPOINTS

Dictionary defining the endpoints for Records REST API.

```python
RECORDS_REST_ENDPOINTS = {
    "docid": {
        "pid_type": DOCUMENT_PID_TYPE,
        "pid_minter": DOCUMENT_PID_MINTER,

>         "pid_fetcher": DOCUMENT_PID_FETCHER,

Reviewer 2: to drop these

        "search_class": DocumentSearch,
        "record_class": Document,

>         "indexer_class": DocumentIndexer,

Reviewer 2: I would leave these and explain that they can be modified if customisation is needed on the Record class behaviour or search class or indexer, but these require some advanced developer experience

        "search_factory_imp": "invenio_app_ils.search_permissions:ils_search_factory",
        "record_loaders": {
            "application/json": "invenio_app_ils.documents.loaders:document_loader"
        },
        "record_serializers": {
            "application/json": "invenio_app_ils.literature.serializers:json_v1_response"
        },
        "search_serializers": {
            "application/json": "invenio_app_ils.literature.serializers:json_v1_search",
            "text/csv": "invenio_app_ils.literature.serializers:csv_v1_search",
        },
        "search_serializers_aliases": {
            "csv": "text/csv",
            "json": "application/json",

>         },

Reviewer 2: to drop

>         "list_route": "/documents/",

Reviewer 2: to drop

>         "item_route": "/documents/<{0}:pid_value>".format(doc_pid_converter()),

Reviewer 2: to drop

>         "default_media_type": "application/json",

Reviewer 2: to drop

        "max_result_window": RECORDS_REST_MAX_RESULT_WINDOW,

>         "error_handlers": {},

Reviewer 2: to drop

        "read_permission_factory_imp": record_read_permission_factory,
        "list_permission_factory_imp": allow_all,
        "create_permission_factory_imp": backoffice_permission,
        "update_permission_factory_imp": backoffice_permission,

>         "delete_permission_factory_imp": backoffice_permission,

Reviewer 2: I would leave this and explain that they are used to implement permission policies for specific endpoints

    },
    # Similar structure as above for other endpoints below
    "pitmid": {...},
    "eitmid": {...},
    "locid": {...},
    "serid": {...},
    "dreqid": {...},
    "vocid": {...},
    "litid": {...},
}
```

For a detailed configuration reference, refer to [invenio-records-rest](https://invenio-records-rest.readthedocs.io/en/latest/).

### RECORDS_REST_SORT_OPTIONS

> Dictionary defining sort options for Records REST endpoints.

Reviewer 2: ```suggestion
Dictionary defining sort options for Records REST endpoints. Sort options ser here will be displayed in the search pages.
```

    document_requests=dict(...),
    series=dict(...),
    items=dict(...),
    eitems=dict(...),
)
```

### ILS_VIEWS_PERMISSIONS_FACTORY = views_permissions_factory

Permissions factory for ILS views to handle all ILS actions.

```python
ILS_VIEWS_PERMISSIONS_FACTORY = views_permissions_factory
```

### ILS_INDEXER_TASK_DELAY

Trigger delay for celery tasks to index referenced records.

```python
ILS_INDEXER_TASK_DELAY = timedelta(seconds=2)
```

## Stats Configuration

For detailed documentation on stats, refer to [invenio-stats](https://github.com/inveniosoftware/invenio-stats).

## ILS Configuration

### ILS_VOCABULARIES (list)

List of available vocabularies.

### ILS_VOCABULARY_SOURCES

Dictionary defining metadata for the vocabularies.

```python
ILS_VOCABULARY_SOURCES = {
    "json": "invenio_app_ils.vocabularies.sources:json_source",
    "opendefinition": (
        "invenio_app_ils.vocabularies.sources:opendefinition_source"
    ),
}
```

### OPENDEFINITION_JSONRESOLVER_HOST

Host URL for opendefinition license for jsonresolvers.

### FILES_REST_PERMISSION_FACTORY

Factory function for defining permissions for file upload.

> ### ILS_RECORDS_EXPLICIT_PERMISSIONS_ENABLED (True/False)

Reviewer 2: To drop

Enable records restrictions by `_access` field.

When enabled, it allows to define explicit permissions for each record to
provide read access to specific users or roles.
When disabled, it will avoid checking for user ids and roles on each search
query and record fetch.

### ILS_LITERATURE_COVER_URLS_BUILDER

Default implementation for building cover urls in document serializer.

> ### ILS_RECORDS_METADATA_NAMESPACES

Reviewer 2: to drop

Namespaces for fields added to the metadata schema.

### ILS_RECORDS_METADATA_EXTENSIONS

> Fields added to the metadata schema.

Reviewer 2: we need an example on how to configure

> ### ILS_PATRON_ANONYMOUS_CLASS and ILS_PATRON_SYSTEM_AGENT_CLASS

Reviewer 2: to drop, I think it is quite unlikely that someone will change it
